### ISSUE TITLE
Implement a Data Source Interface and add the data from the FOSSASIA and Freifunk Api to loklak through the interface

### ISSUE THREAD

**Issue by mariobehling** [loklak/loklak_webclient#206]
Here is an idea draft how data can be added through the interface.
![loklak-datasources-connect-services](https://cloud.githubusercontent.com/assets/1583873/8267842/8eb5f954-176f-11e5-8fa4-7c98fe2f9c9a.png)

This issue is related to https://github.com/loklak/loklak_webclient/issues/205

**Comment by Orbiter** [comment]
the server part which receives this information is described in https://github.com/loklak/loklak_server/issues/35

**Comment by zyzo** [comment]
FOSSASIA/Freifunk API data consist of community json files that follow this specification https://github.com/fossasia/api.fossasia.net/blob/master/specs/1.0.1.json.

To be short, this is the complete list of properties : 

1. api metadata (api version)
2. **community name**
3. metacommunity (bigger community that this community belongs to)
3. **location (city, country, address, coordinates)**
4. **contact info (email, facebook, twitter, phone, irc, ..)**
5. webpage url
6. timeline (lifetime events something like date of birth or other important milestones)
7. **feeds (blog, forum, ics, podcast, wiki, ..)**
8. state : the state of the json file (last-changed, open-closed for public), logo, community description

**Comment by zyzo** [comment]
To integrate FOSSASIA/Freifunk API data into loklak, a few questions come up :

* what kind of adapter / processor need to be written to dump the data into loklak ?
* where's the part of code should I look at for examples of these kinds of adapter ?

Just to clarify things a bit, API data consist of a lot of information about a community, like names, location, services. From the name and location, a community map was built, and now we want to add some filters to it (by country, by tech area..). Sometimes we want to retrieve data directly from services as well (rss, ics feeds..).

**Comment by mariobehling** [comment]
@zyzo 
> what kind of adapter / processor need to be written to dump the data into loklak ?

What is your proposal?

> where's the part of code should I look at for examples of these kinds of adapter ?

We are just getting started with it.

> Just to clarify things a bit, API data consist of a lot of information about a community, like names, location, services. From the name and location, a community map was built, and now we want to add some filters to it (by country, by tech area..). Sometimes we want to retrieve data directly from services as well (rss, ics feeds..).

How can we pump all this data to loklak? Actually we diverted from the original meaning of "API" a bit and also included more to it. In my opinion - let's focus with the API on the core topic and let other services display the data e.g. the Freifunk blog, Loklak and others.

**Comment by mariobehling** [comment]
@sudheesh001 Why has this been closed? Where can I test the implementation?

**Comment by sudheesh001** [comment]
@mariobehling I think this got closed because of the term `Fix #206` which was present in the pull request content https://github.com/loklak/loklak_webclient/pull/259 I've been adding a label `Checked` in case it works and is actually supposed to be closed.

**Comment by zyzo** [comment]
@sudheesh001 I think you meant `#260` ?

**Comment by zyzo** [comment]
This went live http://test.loklak.net/dataConnect

**Comment by zyzo** [comment]
This is what current implementation looks like, with loklak_server new feature https://github.com/loklak/loklak_server/pull/83: 

![screenshot from 2015-07-24 11 08 47](https://cloud.githubusercontent.com/assets/6483446/8867527/b26cf16e-31f4-11e5-9f2b-e6ef4bde5871.png)

TODO : 

* option to modify the datasource directly from web client.
* for geojson source, there is map_type parameter. User should be able to visualize it and modify it as well. This feature requires some more work on the server as the information about map_type is currently not saved.

**Comment by zyzo** [comment]
This feature can be reviewed at http://loklak-client.danghaian.com/dataConnect
(Server : http://loklak-server.danghaian.com)

**Comment by mariobehling** [comment]
@zyzo Nice! loklak-client seems to be down though. Cannot test.

**Comment by zyzo** [comment]
My bad, @mariobehling could you please test again ?

**Comment by mariobehling** [comment]
Login screen for twitter stays empty - cannot login.

![screenshot from 2015-07-24 09 58 06](https://cloud.githubusercontent.com/assets/1583873/8869985/7fe82efc-31ea-11e5-98b0-a83d7891292f.png)

**Comment by zyzo** [comment]
There is an unknown problem with twitter auth indeed, but the datasource interface can currently work without logging in to twitter.

**Comment by mariobehling** [comment]
Ok, a few questions. I could see my connections without logging in.
* Where does the data of http://cmap-fossasia-api.heroku.com/ffGeoJsonp.php come from? How is it generated?
* What does the user have to do in "source type"? How should a user know what to do here?
* What does a user have to do in mapping rules and tweet field?

**Comment by zyzo** [comment]
@mariobehling  : 
* when adding new data source (Add Data Source), the source is pushed via `api/push/geojson.json`. The geojson push servlet do two things : save the messages & the import profile (information about source url, created time, source type, harvesting frequence.. cf. https://github.com/loklak/loklak_server/issues/35). Then the list of import profiles are retrieved via `api/import.json` and displayed in the data source page.

* source_type is a necessary information when pushing messages as geojson. The assumption is that the user has a data file in geojson format, that he wants to import. In order to do that, he must know what is the source_type of the source (Twitter, Fossasia, Freifunk, ..). 

* The mapping rules are optional, but necessary, if the user file isn't compatible with loklak message type (which is twitter's like).

In short the interface is a visual way to construct push query to `api/push/geojson.json` (more about it here http://loklak.org/api.html#geojsonpush). The interface will change to accommodate new servlets `api/push/fossasia.json`, `api/push/openwifimap.json`, ...

**Comment by mariobehling** [comment]
@mariobehling

1. http://cmap-fossasia-api.heroku.com/ffGeoJsonp.php Whose server is it? Who creates this data ffGeoJsonp.php? Why do we need a server running php to offer data for loklak? How can we get the data directly from the communities in github. Think of the use case "I want to add my community to the map" (e.g. https://raw.githubusercontent.com/fossasia/fossasia-communities/master/Singapore/HackerspaceSG).

2. 
> source_type is a necessary information when pushing messages as geojson. The assumption is that the user has a data file in geojson format, that he wants to import. In order to do that, he must know what is the source_type of the source (Twitter, Fossasia, Freifunk, ..). 

I am sorry but you cannot expect the user to know all kinds of formats. Do you remember we had talked about a possible way to do this like mailchimp? We possibly want to support dozens of formats in future. Also FOSSASIA or Freifunk is just a name for nearly the same format. I see a) import, b) user, c) FOSSASIA. What does the user have to do in each "source type"? I do not understand what is import and user.

3. 

> The mapping rules are optional, but necessary, if the user file isn't compatible with loklak message format (which is twitter's like).

This does not answer my question. You answer the question "Are the rules optional or necessary?".

My question is: What does a user have to do in mapping rules and tweet field? What information does he/she have to insert?

**Comment by zyzo** [comment]
> http://cmap-fossasia-api.heroku.com/ffGeoJsonp.php Whose server is it? Who creates this data ffGeoJsonp.php? Why do we need a server running php to offer data for loklak? How can we get the data directly from the communities in github

This is really just a json file that serves http://cmap-fossasia-api.herokuapp.com/ffGeoJson.json, nothing important (It is `.php` to deal with jsonp). And it is used to demo here because at the moment I could only push messages in `geojson` format. Now, there is `/api/push/fossasia.json` that can be used to push data in fossasia api format. The feature just got merged into master. As said I will accommodate this change into the dataConnect interface.

Then, user will be able to add pure fossasia api data.

> I see a) import, b) user, c) FOSSASIA. What does the user have to do in each "source type"? I do not understand what is import and user.

Import and user shoudn't appear, I will remove them. They exist because SourceType was used for a different purpose [SourceType.java](https://github.com/loklak/loklak_server/blob/master/src/org/loklak/harvester/SourceType.java).


> What does a user have to do in mapping rules and tweet field? What information does he/she have to insert?

As explained in http://loklak.org/api.html#geojsonpush, the mapping rule (parameter `map_type`) is used to map fields in original data to loklak message data. For instance, if you have a field called `name`, that should be saved as `screen_name` of the message, then you can make it a rule. More discussion about `map_type` : 

https://github.com/loklak/loklak_server/issues/55#issue-90898076
https://github.com/loklak/loklak_server/issues/55#issuecomment-122170771

**Comment by zyzo** [comment]
Interface to add fossasia api data : 
![screenshot from 2015-07-25 12 20 54](https://cloud.githubusercontent.com/assets/6483446/8888067/e7c61c04-32c7-11e5-8d0b-98654a1778b3.png)

If no format is specified in route parameter `/addConnection/:format`, or if format=geosjon, user will get following form which is a little bit different : 

![screenshot from 2015-07-25 12 21 09](https://cloud.githubusercontent.com/assets/6483446/8888071/2545d2f4-32c8-11e5-91d5-8cba26ea7d5e.png)

**Comment by mariobehling** [comment]
Currently loklak-client is not available.

If the format is not specified, we should have a validation form, that automatically matches the right format. This can be a follow up issue. Regarding the import of API data, please have a look how other services like mailchimp do this: http://kb.mailchimp.com/lists/growth/import-subscribers-to-a-list

**Comment by zyzo** [comment]
I added a setting form to let users change data format. It is opened when clicking on 
the gear icon :  
![screenshot from 2015-07-27 12 38 23](https://cloud.githubusercontent.com/assets/6483446/8899548/17858026-3460-11e5-8aca-067592772122.png)

![screenshot from 2015-07-27 12 38 14](https://cloud.githubusercontent.com/assets/6483446/8899550/1b8921e6-3460-11e5-84f8-4857af696308.png)

Automatically sniffing the right format is a nice idea ! But I think it won't be 100% reliable so we should still let user choose the format her/himself.

**Comment by mariobehling** [comment]
Ok, let's iterate and get this going in the above way first. Thanks!

**Comment by zyzo** [comment]
It is now possible to add Freifunk & FOSSASIA data via the interface. Please refer to : 
* server side :  https://github.com/loklak/loklak_server/pull/77 https://github.com/loklak/loklak_server/pull/84
* client side : #285 #325 #394
